### PR TITLE
Feat/use advanced endpoint search

### PR DESCRIPTION
**Pull request by giacomocamerano**
-switch to advanced search for searching bands (more data, possible to use ExactBandMatch)
- add tests (got carried away and went to 100% coverage on all code)

## Summary by CodeRabbit

- **New Features**
	- Introduced advanced band search functionality with new schemas and methods for enhanced querying.
	- Added new dependencies for improved functionality and testing capabilities.
	- Implemented a function to extract band information from HTML content.

- **Bug Fixes**
	- Updated error handling for band information extraction to improve robustness.

- **Tests**
	- Added comprehensive test suites for `extractBandInfo`, `searchBand`, and `searchBandAdvanced` functions to ensure reliability and correct behavior.
	- Introduced fixtures for testing search queries and responses.

- **Documentation**
	- Enhanced schema definitions with descriptions for better clarity and usability.

### CODE REVIEW

File: src/clients/metalArchives/common.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    htmlString: string
+): { band: string; link: string; id: number } => {
+    const regex =
+        /<a href="(https:\/\/www\.metal-archives\.com\/bands\/[^"]+\/(\d+))">([^<]+)<\/a>/
+
+    const [, link = '', id = '', band = ''] = regex.exec(htmlString) || []
+
+    if (!link || !id || !band) {
+        throw new Error('Failed to parse band information from HTML string')
+    }
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/common.ts:11]: not needed, if regex fails it will most probably fail for all

File: src/clients/metalArchives/common.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+    return { band, link, id: Number.parseInt(id, 10) }
+}
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/common.ts:14]: why do we need this check when we are using regex after? If the destructuring fails, it get caught by the condition

File: src/clients/metalArchives/test/common.test.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { extractBandInfo } from '../common'
+
+describe('extractBandInfo', () => {
+    it('should throw an error if the HTML string does not match the regex pattern (line 10)', () => {
+        const invalidHtmlString = '<p>No link here</p>'
+
+        expect(() => extractBandInfo(invalidHtmlString)).toThrow(
+            'Failed to parse band information from HTML string'
+        )
+    })
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/test/common.test.ts:10]: fixed

File: src/clients/metalArchives/test/common.test.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+            '<a href="https://www.metal-archives.com/bands/Immortal/12345">Immortal</a>'
+
+        const result = extractBandInfo(validHtmlString)
+
+        expect(result).toEqual({
+            band: 'Immortal',
+            link: 'https://www.metal-archives.com/bands/Immortal/12345',
+            id: 12345
+        })
+    })
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/test/common.test.ts:23]: suggest further edge cases and unit tests

File: src/config.ts [hunk 5]
```diff
@@ -28,9 +28,16 @@ const development: Config = {
...
 
-export { options, port }
+export const config = {
+    metalArchives: {
+        searchBandUrl:
+            'https://www.metal-archives.com/search/ajax-band-search/',
+        searchBandAdvancedUrl:
+            'https://www.metal-archives.com/search/ajax-advanced/searching/bands/'
+    }
+}
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/config.ts:43]: will do in later PR

File: src/clients/metalArchives/schema/searchBandAdvanced.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    ),
+    status: Type.Optional(
+        Type.Array(statusSchema, {
+            description: 'Array of status codes with their meanings'
+        })
+    ),
+
+    themes: Type.Optional(
+        Type.String({ description: 'Search by themes', default: '*' })
+    ),
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/schema/searchBandAdvanced.ts:67]: used this because I always want the field to be populated. If it is empty, themes won't be returned in response from MA

File: src/clients/metalArchives/schema/searchBandAdvanced.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    Type.String({
+        description: 'Complete band link for MA, including name, link, id'
+    }),
+    Type.String({ description: 'Genre' }),
+    Type.String({ description: 'Country' }),
+    Type.String({ description: 'Location (city/region)' }),
+    Type.String({ description: 'themes' }),
+    Type.String({ description: 'Year of band founding' }),
+    Type.String({ description: 'Label name' })
+])
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/schema/searchBandAdvanced.ts:101]: I would but that is what is returned by MA

File: src/clients/metalArchives/test/common.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/test/common.test.ts:23]: added to upcoming commit, thank you

File: src/clients/metalArchives/schema/searchBandAdvanced.ts [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+            description: 'Array of status codes with their meanings'
+        })
+    ),
+
+    themes: Type.Optional(
+        Type.String({ description: 'Search by themes', default: '*' })
+    ),
+    location: Type.Optional(
+        Type.String({ description: 'Search by themes', default: '*' })
+    ),
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/schema/searchBandAdvanced.ts:70]: fixed in upcoming commit

File: src/clients/metalArchives/searchBandAdvanced.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        yearCreationTo?: number
+        bandNotes?: string
+        status?: Status[]
+        themes?: string
+        location?: string
+        bandLabelName?: string
+    }
+) => {
+    const {
+        status = [1, 2, 3, 4, 5, 6],
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/searchBandAdvanced.ts]: fixed in upcoming commit

File: src/clients/metalArchives/searchBandAdvanced.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+                id,
+                genre,
+                country,
+                location,
+                themes,
+                year,
+                label
+            } satisfies BandSearchResult
+        }
+    )
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/searchBandAdvanced.ts:112]: this is not needed: if extractBandInfo is not able to return data, an error is thrown inside extractBandInfo, no need to catch and rethrow here. Wdyt?

File: src/clients/metalArchives/schema/searchBandAdvanced.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+            description: 'Year of creation from',
+            examples: [2000, 2010]
+        })
+    ),
+    yearCreationTo: Type.Optional(
+        Type.Number({
+            description: 'Year of creation to',
+            examples: [2000, 2010]
+        })
+    ),
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/schema/searchBandAdvanced.ts:61]: nice catch. No need for runtime validation, if wrong will just return empty result. Have added the validation though

File: src/clients/metalArchives/schema/searchBandAdvanced.ts [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+export const searchBandAdvancedResponseSchema = Type.Object({
+    error: Type.Optional(Type.String()),
+    iTotalRecords: Type.Number(),
+    iTotalDisplayRecords: Type.Number(),
+    sEcho: Type.Number(),
+    aaData: Type.Array(searchBandAdvancedEntrySchema)
+})
+export type SearchBandAdvancedResponse = Static<
+    typeof searchBandAdvancedResponseSchema
+>
```
[the remark this answers is not in the document]
Comment (giacomocamerano) [reply src/clients/metalArchives/schema/searchBandAdvanced.ts:125]: not under my control, external endpoint
